File: riceCooker/src/test/java/org/riceCooker/ut/RiceCookerTest.java
package org.riceCooker.ut;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.riceCooker.classes.RiceCooker;

import static org.junit.jupiter.api.Assertions.*;
import static org.riceCooker.utils.Constants.getMessage;

class RiceCookerTest {
    private RiceCooker rc = new RiceCooker();

    @BeforeEach
    void newRiceCooker() {
        rc = new RiceCooker();
    }

    @Test
    void power_on_not_plugged_ko() {
        assertThrows(RuntimeException.class, () -> rc.togglePower());
        try {
            rc.togglePower();
        } catch (RuntimeException re) {
            assertEquals(re.getMessage(), getMessage("error:power-not-plugged"));
        }
    }

    @Test
    void plug_ok() {
        assertDoesNotThrow(() -> rc.togglePlug());
    }

    @Test
    void power_on_plugged_ok() {
        rc.togglePlug();
        assertTrue(rc.getPlug());
        assertDoesNotThrow(() -> rc.togglePower());
    }

    @Test
    void add_water_ok() {
        final int waterToAdd = 10;
        assertDoesNotThrow(() -> rc.addWater(waterToAdd));
    }

    @Test
    void add_water_ko() {
        final int waterToAdd = 11;
        assertThrows(RuntimeException.class, () -> rc.addWater(waterToAdd));
        try {
            rc.addWater(waterToAdd);
        } catch (RuntimeException re) {
            assertEquals(re.getMessage(), getMessage("error:max-capacity"));
        }
    }

    @Test
    void remove_water_ok() {
        final int waterToAdd = 6;
        final int waterToRemove = 5;
        final int expectedRest = 1;

        rc.addWater(waterToAdd);
        assertDoesNotThrow(() -> rc.removeWater(waterToRemove));
        assertEquals(expectedRest, rc.getWaterCupNumber());
    }


    @Test
    void remove_water_ko() {
        final int waterToRemove = 1;
        assertThrows(RuntimeException.class, () -> rc.removeWater(waterToRemove));
        try {
            rc.addWater(waterToRemove);
        } catch (RuntimeException re) {
            assertEquals(re.getMessage(), getMessage("error:not-enough-water"));
        }
    }
    @Test
    void add_rice_ok() {
        final int riceToAdd = 10;
        assertDoesNotThrow(() -> rc.addRice(riceToAdd));
    }

    @Test
    void add_rice_ko() {
        final int riceToAdd = 11;
        assertThrows(RuntimeException.class, () -> rc.addRice(riceToAdd));
        try {
            rc.addRice(riceToAdd);
        } catch (RuntimeException re) {
            assertEquals(re.getMessage(), getMessage("error:max-capacity"));
        }
    }

    @Test
    void remove_rice_ok() {
        final int riceToAdd = 6;
        final int riceToRemove = 5;
        final int expectedRest = 1;

        rc.addRice(riceToAdd);
        assertDoesNotThrow(() -> rc.discardRice(riceToRemove));
        assertEquals(expectedRest, rc.getRiceCupNumber());
    }


    @Test
    void remove_rice_ko() {
        final int riceToRemove = 1;
        assertThrows(RuntimeException.class, () -> rc.discardRice(riceToRemove));
        try {
            rc.discardRice(riceToRemove);
        } catch (RuntimeException re) {
            assertEquals(re.getMessage(), getMessage("error:not-enough-rice"));
        }
    }
}

File: tests/rice_cooker_test.go
package tests

import (
	"testing"

	"co.hei.ricecooker/internal/classes"
	"co.hei.ricecooker/packages/constants"
)

var rc = classes.NewRiceCooker()

func TestPowerOnNotPluggedKo(t *testing.T) {
	err := rc.TogglePower()
	expectedErrorMessage := constants.Messages["error:power-not-plugged"]
	if err == nil {
		t.Error("Should return error 'cause rice cooker is not already plugged.")
	} else if err.Error() != expectedErrorMessage {
		t.Errorf("Expected error %s but got %s", expectedErrorMessage, err.Error())
	}
}

func TestPlugOk(t *testing.T) {
	err := rc.TogglePlug()
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}

	isPlug := rc.GetPlug()
	if !isPlug {
		t.Error("Plug variable should be true but got false")
	}
}

func TestPowerOnPluggedOk(t *testing.T) {
	err := rc.TogglePower()
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}
}

func TestAddWaterOk(t *testing.T) {
	validWaterCup := 5
	err := rc.AddWater(validWaterCup)
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}
	currentWaterCup := rc.GetWaterCupNumber()
	if validWaterCup != currentWaterCup {
		t.Errorf("rc.GetWaterCupNumber() return %d, expected %d", currentWaterCup, validWaterCup)
	}
}

func TestAddWaterKo(t *testing.T) {
	notValidaWaterCup := 15
	err := rc.AddWater(notValidaWaterCup)
	expectedError := constants.Messages["error:max-capacity"]

	if err == nil {
		t.Errorf("Should return error")
	} else if expectedError != err.Error() {
		t.Errorf("Expected error %s but got %s", expectedError, err.Error())
	}
}

func TestAddRiceOk(t *testing.T) {
	validRiceCup := 5
	err := rc.AddRice(validRiceCup)
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}
	currentRiceCup := rc.GetRiceCupNumber()
	if validRiceCup != currentRiceCup {
		t.Errorf("rc.GetRiceCupNumber() return %d, expected %d", currentRiceCup, validRiceCup)
	}
}

func TestAddRiceKo(t *testing.T) {
	notValidaRiceCup := 15
	err := rc.AddRice(notValidaRiceCup)
	expectedError := constants.Messages["error:max-capacity"]

	if err == nil {
		t.Errorf("Should return error")
	} else if expectedError != err.Error() {
		t.Errorf("Expected error %s but got %s", expectedError, err.Error())
	}
}

func TestRemoveWaterOk(t *testing.T) {
	validWaterCup := 1
	err := rc.RemoveWater(validWaterCup)
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}
}
func TestRemoveWaterKo(t *testing.T) {
	notValidWaterCup := 10
	err := rc.RemoveWater(notValidWaterCup)
	expectedError := constants.Messages["error:not-enough-water"]

	if err == nil {
		t.Errorf("Should return error")
	} else if expectedError != err.Error() {
		t.Errorf("Expected error %s but got %s", expectedError, err.Error())
	}
}

func TestDiscardRiceOk(t *testing.T) {
	validRiceCup := 1
	err := rc.DiscardRice(validRiceCup)
	if err != nil {
		t.Errorf("Shouldn't return error but got %s", err.Error())
	}
}
func TestDiscardRiceKo(t *testing.T) {
	notValidRiceCup := 10
	err := rc.DiscardRice(notValidRiceCup)
	expectedError := constants.Messages["error:not-enough-rice"]

	if err == nil {
		t.Errorf("Should return error")
	} else if expectedError != err.Error() {
		t.Errorf("Expected error %s but got %s", expectedError, err.Error())
	}
}
